proc_macro/bridge: Eagerly pass TokenStream contents into the client

mystor opened the pull request:
This change could improve performance, especially in cases like with the
cross-thread backend which have expensive RPC. It handles this by moving
more of the logic for TokenStream into the client, storing a
`Rc<Vec<TokenTree>>` directly, rather than converting to/from the
compiler representation over RPC for every change.

As this eagerly decomposes interpolated AST fragments, this ended up
changing the output of `TokenStream::to_string()`. To keep up
compatibility for now, formatting is still performed using RPC and
`rustc_ast_pretty`, however in the future we may look into decoupling
proc-macro formatting from `rustc_ast_pretty`, and doing it in the
client.

One side-effect of this was that the workaround for the
procedural-masquerade crate had to be updated to work with the new
approach, as we couldn't depend on the AST fragment printer to format
the input. To do this, a real `,` token is now inserted when passing the
enum to a proc-macro rather than depending on the pretty-printing
behaviour. From local tests it appears that the new output should be
compatible enough to work with older versions of procedural-masquerade.

The new conversion between the proc-macro and compiler representations
of a `TokenStream` is now handled by the `RpcContext` type provided by
the proc-macro server. This type is a distinct value which will be
fetched and stored by the server before commands are run, such that the
decoders can have access to relevant state. In the future, we may also
change the symbol handling code to work off of `RpcContext` rather than
being static methods on the `Server` trait.

rust-highfive commented:
r? @petrochenkov

(rust-highfive has picked a reviewer for you, use r? to override)

mystor commented:
Posting this change as a work-in-progress change, which is not ready to merge. I think we'll want to run this through crater before we move too much further with it (assuming it passes CI), to make sure it doesn't cause breakage in the ecosystem.

r? @eddyb

mystor commented:
Another potential concern here is that the code for serializing and deserializing `TokenStream` with this patch is written in a recursive manner, as is the stringification/formatting code. I think the formatting code previously (from `rustc_ast_pretty`) was also recursive, but this could theoretically exhaust the stack for sufficiently nested `TokenTree`s. 

That being said, it should be possible to rework things to avoid using the stack in the future, but keeping it this way for now makes everything much simpler.

eddyb commented:
Might be too early for this, but curious anyway:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 7a64bb64f92fa2c6741c3b4938804514e46b0097 with merge 0c7f4114a5582dfa18524643326a0e87bc6d62f3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"7a64bb64f92fa2c6741c3b4938804514e46b0097","merge_sha":"0c7f4114a5582dfa18524643326a0e87bc6d62f3"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8178875238?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7a64bb64f92fa2c6741c3b4938804514e46b0097/checks/8178875238/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
........................................................................................ 10560/13475
........................................................................................ 10648/13475
............................................iiiii....i...i.i............................ 10736/13475
........................................................................................ 10824/13475
.........F.F.i.......................................................................... 10912/13475
........................................................................................ 11088/13475
........................................................................................ 11176/13475
........................................................................................ 11264/13475
........................................................................................ 11352/13475
---
---- [ui] src/test/ui/rfc-2565-param-attrs/param-attrs-pretty.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-2565-param-attrs/param-attrs-pretty.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2565-param-attrs/param-attrs-pretty" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2565-param-attrs/param-attrs-pretty/auxiliary"
stdout: none
--- stderr -------------------------------
error: custom attribute panicked
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
   |
   |
LL | #[attr_extern]
   |
   = help: message: assertion failed: `(left == right)`
   = help: message: assertion failed: `(left == right)`
             left: `"extern \"C\" { fn ffi(#[a1] arg1 : i32, #[a2]...) ; }"`,
            right: `"extern \"C\" { fn ffi(#[a1] arg1 : i32, #[a2] ...) ; }"`

error: custom attribute panicked
   |
   |
LL | #[attr_extern_cvar]
   |
   = help: message: assertion failed: `(left == right)`
   = help: message: assertion failed: `(left == right)`
             left: `"unsafe extern \"C\" fn cvar(arg1 : i32, #[a1] mut args :...) {}"`,
            right: `"unsafe extern \"C\" fn cvar(arg1 : i32, #[a1] mut args : ...) {}"`

error: custom attribute panicked
   |
LL | #[attr_alias]
   | ^^^^^^^^^^^^^
   |
   |
   = help: message: assertion failed: `(left == right)`
             left: `"type Alias = fn(#[a1] u8, #[a2]...) ;"`,
            right: `"type Alias = fn(#[a1] u8, #[a2] ...) ;"`

error: custom attribute panicked
   |
   |
LL | #[attr_free]
   |
   = help: message: assertion failed: `(left == right)`
   = help: message: assertion failed: `(left == right)`
             left: `"fn free(#[a1] arg1 : u8) { let lam = |#[a2] W(x), #[a3] y | () ; }"`,
            right: `"fn free(#[a1] arg1 : u8) { let lam = | #[a2] W(x), #[a3] y | () ; }"`

error: custom attribute panicked
   |
LL |     #[attr_trait_4]
   |     ^^^^^^^^^^^^^^^
   |
   |
   = help: message: assertion failed: `(left == right)`
             left: `"fn trait4 < 'a > (#[a1] self : Box < Self >, #[a2] arg1 : u8, #[a3] Vec < u8 >) ;"`,
            right: `"fn trait4 < 'a >\n(#[a1] self : Box < Self >, #[a2] arg1 : u8, #[a3] Vec < u8 >) ;"`
error: aborting due to 5 previous errors
------------------------------------------



---- [ui] src/test/ui/rfc-2565-param-attrs/issue-64682-dropping-first-attrs-in-impl-fns.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/rfc-2565-param-attrs/issue-64682-dropping-first-attrs-in-impl-fns.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2565-param-attrs/issue-64682-dropping-first-attrs-in-impl-fns" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/rfc-2565-param-attrs/issue-64682-dropping-first-attrs-in-impl-fns/auxiliary"
stdout: none
--- stderr -------------------------------
error: custom attribute panicked
   |
   |
LL | #[rename_params(send_help)]
   |
   = help: message: assertion failed: `(left == right)`
   = help: message: assertion failed: `(left == right)`
             left: `"impl Foo { fn hello(#[angery(true)] a : i32, #[a2] b : i32, #[what = \"how\"] c : u32) {} fn hello2(#[a1] #[a2] a : i32, #[what = \"how\"] b : i32, #[angery(true)] c : u32) {} fn hello_self(#[a1] #[a2] & self, #[a1] #[a2] a : i32, #[what = \"how\"] b : i32, #[angery(true)] c : u32) {} }"`,
            right: `"impl Foo\n{\n    fn hello(#[angery(true)] a : i32, #[a2] b : i32, #[what = \"how\"] c : u32)\n    {} fn\n    hello2(#[a1] #[a2] a : i32, #[what = \"how\"] b : i32, #[angery(true)] c :\n    u32) {} fn\n    hello_self(#[a1] #[a2] & self, #[a1] #[a2] a : i32, #[what = \"how\"] b :\n    i32, #[angery(true)] c : u32) {}\n}"`
error: aborting due to previous error
------------------------------------------


```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/8179481597?check_suite_focus=true)
Build commit: 0c7f4114a5582dfa18524643326a0e87bc6d62f3 (`0c7f4114a5582dfa18524643326a0e87bc6d62f3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/8179481597?check_suite_focus=true"},"merge_sha":"0c7f4114a5582dfa18524643326a0e87bc6d62f3"} -->

rust-timer commented:
Queued 0c7f4114a5582dfa18524643326a0e87bc6d62f3 with parent a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318&end=0c7f4114a5582dfa18524643326a0e87bc6d62f3).

mystor commented:
Looks like I missed 2 proc macros in tests with changes to their output formats. Fortunately shouldn't be hard to update them.

mystor committed:
proc_macro/bridge: Eagerly pass TokenStream contents into the client

This change could improve performance, especially in cases like with the
cross-thread backend which have expensive RPC. It handles this by moving
more of the logic for TokenStream into the client, storing a
`Rc<Vec<TokenTree>>` directly, rather than converting to/from the
compiler representation over RPC for every change.

As this eagerly decomposes interpolated AST fragments, this ended up
changing the output of `TokenStream::to_string()`. To keep up
compatibility for now, formatting is still performed using RPC and
`rustc_ast_pretty`, however in the future we may look into decoupling
proc-macro formatting from `rustc_ast_pretty`, and doing it in the
client.

One side-effect of this was that the workaround for the
procedural-masquerade crate had to be updated to work with the new
approach, as we couldn't depend on the AST fragment printer to format
the input. To do this, a real `,` token is now inserted when passing the
enum to a proc-macro rather than depending on the pretty-printing
behaviour. From local tests it appears that the new output should be
compatible enough to work with older versions of procedural-masquerade.

The new conversion between the proc-macro and compiler representations
of a `TokenStream` is now handled by the `RpcContext` type provided by
the proc-macro server. This type is a distinct value which will be
fetched and stored by the server before commands are run, such that the
decoders can have access to relevant state. In the future, we may also
change the symbol handling code to work off of `RpcContext` rather than
being static methods on the `Server` trait.

mystor commented:
It occurred to me that pretty-printing a large `TokenStream` might be rare enough nowadays that we can afford sending the entire TokenStream over IPC just to pretty-print it, at least as a stopgap, to avoid some of the pretty-printing changes from also happening in this patch.

I still have the other changes to do it more efficiently, but we can potentially leave that for a follow-up if the perf on this approach is OK.

This still requires the changes to the pretty-printing hack for `procedural-masquerade` because the interpolated node can't be used anymore.

mystor commented:
cc #73933 (@Aaron1011 @petrochenkov), as that issue tracks removing the hack completely.

rust-timer commented:
Finished benchmarking commit (0c7f4114a5582dfa18524643326a0e87bc6d62f3): [comparison URL](https://perf.rust-lang.org/compare.html?start=a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318&end=0c7f4114a5582dfa18524643326a0e87bc6d62f3&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318&end=0c7f4114a5582dfa18524643326a0e87bc6d62f3&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 1.2%     | [0.3%, 4.3%] | 35        |
| Regressions ❌ <br /> (secondary) | 0.9%     | [0.7%, 1.0%] | 4         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -15.9%   | [-43.7%, -1.0%] | 11        |
| All ❌✅ (primary) | 1.2%     | [0.3%, 4.3%] | 35        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318&end=0c7f4114a5582dfa18524643326a0e87bc6d62f3&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 4.3%     | [4.3%, 4.3%] | 1         |
| Regressions ❌ <br /> (secondary) | 3.0%     | [2.6%, 3.5%] | 2         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.0%    | [-3.9%, -2.1%] | 2         |
| All ❌✅ (primary) | 4.3%     | [4.3%, 4.3%] | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a2cdcb3fea2baae5d20eabaa412e0d2f5b98c318&end=0c7f4114a5582dfa18524643326a0e87bc6d62f3&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 3.0%     | [1.4%, 6.7%] | 8         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -12.5%   | [-31.7%, -2.2%] | 10        |
| All ❌✅ (primary) | 3.0%     | [1.4%, 6.7%] | 8         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

bors commented:
:umbrella: The latest upstream changes (presumably #101986) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@mystor any updates on this?

mystor commented:
No, I unfortunately haven't had the time to dig into fixing this patch up & getting it ready to be looked into more.

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
